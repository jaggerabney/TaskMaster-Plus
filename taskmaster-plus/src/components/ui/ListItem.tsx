import { Task } from "@/contexts/ListContext";
import Card from "./Card";
import TaskItem from "./TaskItem";

export interface ListItemType {
  id: number;
  title: string;
  tasks: Task[];
  visible: boolean;
  onClick: () => void;
}

const ListItem: React.FC<ListItemType> = ({
  id,
  title,
  tasks,
  visible,
  onClick
}) => {
  if (visible) {
    return (
      <Card key={id}>
        <h2 className="text-2xl font-bold ">{title}</h2>
        {tasks.length > 0 &&
          tasks.map((task) => (
            <TaskItem
              key={task.id}
              id={task.id}
              listId={task.listId}
              title={task.title}
              completed={task.completed}
              dueDate={task.dueDate}
              description={task.description}
              rrule={task.rrule}
            />
          ))}
        {tasks.length == 0 && (
          <div
            className="self-center underline text-redNCS cursor-pointer"
            onClick={onClick}
          >
            Create new task
          </div>
        )}
      </Card>
    );
  }
};

export default ListItem;
