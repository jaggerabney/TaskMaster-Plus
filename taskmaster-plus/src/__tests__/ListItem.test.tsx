import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import ListItem from "@/components/ui/ListItem";

describe("List item component", () => {
  const mockListItemProps = {
    id: 1,
    title: "Untitled list",
    tasks: [],
    visible: true,
    onClick: () => {}
  };

  it("renders correctly", () => {
    const listItem = render(<ListItem {...mockListItemProps} />);

    // no act

    expect(listItem).toMatchSnapshot();
  });

  it("is visible when visible = true", () => {
    render(<ListItem {...mockListItemProps} />);
    const listItem = screen.queryByText(mockListItemProps.title);

    // no act

    expect(listItem).not.toBeNull();
  });

  it("isn't visible when visible = false", () => {
    const invisibleListItemProps = {
      ...mockListItemProps,
      visible: false
    };

    render(<ListItem {...invisibleListItemProps} />);
    const listItem = screen.queryByText(mockListItemProps.title, {});

    expect(listItem).toBeNull();
  });

  it('displays the "Create new task" link when there are no tasks', () => {
    render(<ListItem {...mockListItemProps} />);
    const createNewTaskLink = screen.queryByText(/create new task/i);

    expect(createNewTaskLink).not.toBeNull();
  });

  it('hides the "Create new task" link when there are tasks', () => {
    const listItemWithTasksProps = {
      ...mockListItemProps,
      tasks: [
        {
          id: 2,
          listId: 1,
          title: "Test task",
          completed: false,
          dueDate: new Date(2030, 0, 1),
          description: "Test description"
        }
      ]
    };

    render(<ListItem {...listItemWithTasksProps} />);
    const createNewTaskLink = screen.queryByText(/create new task/i);

    expect(createNewTaskLink).toBeNull();
  });
});
