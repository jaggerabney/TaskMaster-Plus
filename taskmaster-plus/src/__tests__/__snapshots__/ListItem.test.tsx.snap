// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List item component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-cosmicLatte rounded p-8 shadow-md flex flex-col gap-4 max-w-md self-start"
      >
        <h2
          class="text-2xl font-bold "
        >
          Untitled list
        </h2>
        <div
          class="self-center underline text-redNCS cursor-pointer"
        >
          Create new task
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-cosmicLatte rounded p-8 shadow-md flex flex-col gap-4 max-w-md self-start"
    >
      <h2
        class="text-2xl font-bold "
      >
        Untitled list
      </h2>
      <div
        class="self-center underline text-redNCS cursor-pointer"
      >
        Create new task
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
